### ISSUE TITLE
[BUG] "Methods of generative fun" notebook crashes on simulate (TFP issue).

### ISSUE THREAD

**Issue by derifatives** [genjax-community/genjax#1263]
**Describe the bug**
In a new conda environment as of today (2024-08-08), the notebook https://github.com/probcomp/genjax/blob/2489b60a408cfc655b95f5e186999d2da19770bc/notebooks/cookbook/basics/methods_of_generative_fun.ipynb#L101 doesn't run properly.

I believe this is a TFP issue (see discussion at bottom), which I'm investigating.

**To Reproduce**
```
import jax
from genjax import ChoiceMapBuilder as C
from genjax import GenericProblem, bernoulli, beta, gen, pretty
from genjax.incremental import Diff, NoChange, UnknownChange
from jax import jit

key = jax.random.PRNGKey(0)
pretty()

# Define a generative function
@gen
def beta_bernoulli_process(u):
    p = beta(1.0, u) @ "p"
    v = bernoulli(p) @ "v"
    return 2 * v

key, subkey = jax.random.split(key)
trace = jax.jit(beta_bernoulli_process.simulate)(subkey, (0.5,))
```

Produces this stack trace:
```
---------------------------------------------------------------------------
OverflowError                             Traceback (most recent call last)
Cell In[4], line 3
      1 # key, subkey = jax.random.split(key)
      2 subkey = jax.random.PRNGKey(18)
----> 3 trace = jax.jit(beta_bernoulli_process.simulate)(subkey, (0.5,))

    [... skipping hidden 12 frame]

File <@beartype(genjax._src.generative_functions.static.StaticGenerativeFunction.simulate) at 0x7fbfdf210400>:55, in simulate(__beartype_object_140462475975040, __beartype_get_violation, __beartype_conf, __beartype_object_93892713132864, __beartype_func, *args, **kwargs)

File [/opt/conda/envs/genjax/lib/python3.11/site-packages/genjax/_src/generative_functions/static.py:524](http://localhost:8080/opt/conda/envs/genjax/lib/python3.11/site-packages/genjax/_src/generative_functions/static.py#line=523), in StaticGenerativeFunction.simulate(self, key, args)
    514 @GenerativeFunction.gfi_boundary
    515 @typecheck
    516 def simulate(
   (...)
    519     args: Tuple,
    520 ) -> StaticTrace:
    521     syntax_sugar_handled = push_trace_overload_stack(
    522         handler_trace_with_static, self.source
    523     )
--> 524     (args, retval, address_visitor, address_traces, score) = simulate_transform(
    525         syntax_sugar_handled
    526     )(key, args)
    527     return StaticTrace(
    528         self,
    529         args,
   (...)
    533         score,
    534     )

File [/opt/conda/envs/genjax/lib/python3.11/site-packages/genjax/_src/generative_functions/static.py:274](http://localhost:8080/opt/conda/envs/genjax/lib/python3.11/site-packages/genjax/_src/generative_functions/static.py#line=273), in simulate_transform.<locals>.wrapper(key, args)
    271 @functools.wraps(source_fn)
    272 def wrapper(key, args):
    273     stateful_handler = SimulateHandler(key)
--> 274     retval = forward(source_fn)(stateful_handler, *args)
    275     (
    276         address_visitor,
    277         address_traces,
    278         score,
    279     ) = stateful_handler.yield_state()
    280     return (
    281         args,
    282         retval,
   (...)
    285         score,
    286     )

File <@beartype(genjax._src.core.interpreters.forward.forward.wrapped) at 0x7fbf2a5eade0>:32, in wrapped(__beartype_object_93892699113776, __beartype_get_violation, __beartype_conf, __beartype_func, *args, **kwargs)

File [/opt/conda/envs/genjax/lib/python3.11/site-packages/genjax/_src/core/interpreters/forward.py:257](http://localhost:8080/opt/conda/envs/genjax/lib/python3.11/site-packages/genjax/_src/core/interpreters/forward.py#line=256), in forward.<locals>.wrapped(stateful_handler, *args)
    253 @functools.wraps(f)
    254 @typecheck
    255 def wrapped(stateful_handler: StatefulHandler, *args):
    256     interpreter = ForwardInterpreter()
--> 257     return interpreter.run_interpreter(
    258         stateful_handler,
    259         f,
    260         *args,
    261     )

File [/opt/conda/envs/genjax/lib/python3.11/site-packages/genjax/_src/core/interpreters/forward.py:240](http://localhost:8080/opt/conda/envs/genjax/lib/python3.11/site-packages/genjax/_src/core/interpreters/forward.py#line=239), in ForwardInterpreter.run_interpreter(self, stateful_handler, fn, *args, **kwargs)
    237 def _inner(*args):
    238     return fn(*args, **kwargs)
--> 240 _closed_jaxpr, (flat_args, _, out_tree) = stage(_inner)(*args)
    241 _jaxpr, consts = _closed_jaxpr.jaxpr, _closed_jaxpr.literals
    242 flat_out = self._eval_jaxpr_forward(
    243     stateful_handler,
    244     _jaxpr,
    245     consts,
    246     flat_args,
    247 )

File [/opt/conda/envs/genjax/lib/python3.11/site-packages/genjax/_src/core/interpreters/staging.py:124](http://localhost:8080/opt/conda/envs/genjax/lib/python3.11/site-packages/genjax/_src/core/interpreters/staging.py#line=123), in stage.<locals>.wrapped(*args, **kwargs)
    122 flat_fun, out_tree = api_util.flatten_fun_nokwargs(fun, in_tree)
    123 flat_avals = safe_map(get_shaped_aval, flat_args)
--> 124 typed_jaxpr = cached_stage_dynamic(flat_fun, tuple(flat_avals))
    125 return typed_jaxpr, (flat_args, in_tree, out_tree)

    [... skipping hidden 1 frame]

File [/opt/conda/envs/genjax/lib/python3.11/site-packages/genjax/_src/core/interpreters/staging.py:111](http://localhost:8080/opt/conda/envs/genjax/lib/python3.11/site-packages/genjax/_src/core/interpreters/staging.py#line=110), in cached_stage_dynamic(flat_fun, in_avals)
    109 @lu.cache
    110 def cached_stage_dynamic(flat_fun, in_avals):
--> 111     jaxpr, _, consts = pe.trace_to_jaxpr_dynamic(flat_fun, in_avals)
    112     typed_jaxpr = jc.ClosedJaxpr(jaxpr, consts)
    113     return typed_jaxpr

    [... skipping hidden 5 frame]

File [/opt/conda/envs/genjax/lib/python3.11/site-packages/genjax/_src/core/interpreters/forward.py:238](http://localhost:8080/opt/conda/envs/genjax/lib/python3.11/site-packages/genjax/_src/core/interpreters/forward.py#line=237), in ForwardInterpreter.run_interpreter.<locals>._inner(*args)
    237 def _inner(*args):
--> 238     return fn(*args, **kwargs)

File [/opt/conda/envs/genjax/lib/python3.11/site-packages/genjax/_src/core/generative/core.py:1602](http://localhost:8080/opt/conda/envs/genjax/lib/python3.11/site-packages/genjax/_src/core/generative/core.py#line=1601), in push_trace_overload_stack.<locals>.wrapped(*args)
   1600 def wrapped(*args):
   1601     GLOBAL_TRACE_OP_HANDLER_STACK.append(handler)
-> 1602     ret = fn(*args)
   1603     GLOBAL_TRACE_OP_HANDLER_STACK.pop()
   1604     return ret

File [/opt/conda/envs/genjax/lib/python3.11/site-packages/genjax/_src/core/pytree.py:497](http://localhost:8080/opt/conda/envs/genjax/lib/python3.11/site-packages/genjax/_src/core/pytree.py#line=496), in Closure.__call__(self, *args, **kwargs)
    496 def __call__(self, *args, **kwargs):
--> 497     return self.fn(*self.dyn_args, *args, **kwargs)

Cell In[3], line 14, in beta_bernoulli_process(u)
     12 @gen
     13 def beta_bernoulli_process(u):
---> 14     p = beta(1.0, u) @ "p"
     15     v = bernoulli(p) @ "v"
     16     return 2 * v

File [/opt/conda/envs/genjax/lib/python3.11/site-packages/genjax/_src/core/generative/core.py:1658](http://localhost:8080/opt/conda/envs/genjax/lib/python3.11/site-packages/genjax/_src/core/generative/core.py#line=1657), in GenerativeFunctionClosure.__matmul__(self, addr)
   1652     return handle_off_trace_stack(
   1653         addr,
   1654         maybe_kwarged_gen_fn,
   1655         (self.args, self.kwargs),
   1656     )
   1657 else:
-> 1658     return handle_off_trace_stack(
   1659         addr,
   1660         self.gen_fn,
   1661         self.args,
   1662     )

File [/opt/conda/envs/genjax/lib/python3.11/site-packages/genjax/_src/core/generative/core.py:1592](http://localhost:8080/opt/conda/envs/genjax/lib/python3.11/site-packages/genjax/_src/core/generative/core.py#line=1591), in handle_off_trace_stack(addr, gen_fn, args)
   1590 if GLOBAL_TRACE_OP_HANDLER_STACK:
   1591     handler = GLOBAL_TRACE_OP_HANDLER_STACK[-1]
-> 1592     return handler(addr, gen_fn, args)
   1593 else:
   1594     raise Exception(
   1595         "Attempting to invoke trace outside of a tracing context.\nIf you want to invoke the generative function closure, and recieve a return value,\ninvoke it with a key."
   1596     )

File <@beartype(genjax._src.generative_functions.static.handler_trace_with_static) at 0x7fbfdf210040>:91, in handler_trace_with_static(__beartype_getrandbits, __beartype_get_violation, __beartype_conf, __beartype_object_93892699190048, __beartype_func, *args, **kwargs)

File [/opt/conda/envs/genjax/lib/python3.11/site-packages/genjax/_src/generative_functions/static.py:467](http://localhost:8080/opt/conda/envs/genjax/lib/python3.11/site-packages/genjax/_src/generative_functions/static.py#line=466), in handler_trace_with_static(addr, gen_fn, args)
    461 @typecheck
    462 def handler_trace_with_static(
    463     addr: StaticAddressComponent | StaticAddress,
    464     gen_fn: GenerativeFunction,
    465     args: Tuple,
    466 ):
--> 467     return trace(addr if isinstance(addr, tuple) else (addr,), gen_fn, args)

File <@beartype(genjax._src.generative_functions.static.trace) at 0x7fbfdf201d00>:89, in trace(__beartype_getrandbits, __beartype_get_violation, __beartype_conf, __beartype_object_93892699190048, __beartype_func, *args, **kwargs)

File [/opt/conda/envs/genjax/lib/python3.11/site-packages/genjax/_src/generative_functions/static.py:178](http://localhost:8080/opt/conda/envs/genjax/lib/python3.11/site-packages/genjax/_src/generative_functions/static.py#line=177), in trace(addr, gen_fn, args)
    170 """Invoke a generative function, binding its generative semantics with the current
    171 caller.
    172 
   (...)
    175     gen_fn: A generative function invoked as a callee of `StaticGenerativeFunction`.
    176 """
    177 addr = Pytree.tree_const(addr)
--> 178 return initial_style_bind(trace_p)(_abstract_gen_fn_call)(
    179     addr,
    180     gen_fn,
    181     args,
    182 )

File [/opt/conda/envs/genjax/lib/python3.11/site-packages/genjax/_src/core/interpreters/forward.py:121](http://localhost:8080/opt/conda/envs/genjax/lib/python3.11/site-packages/genjax/_src/core/interpreters/forward.py#line=120), in initial_style_bind.<locals>.bind.<locals>.wrapped(*args, **kwargs)
    119 def wrapped(*args, **kwargs):
    120     """Runs a function and binds it to a call primitive."""
--> 121     _jaxpr, (flat_args, in_tree, out_tree) = stage(f)(*args, **kwargs)
    122     outs = prim.bind(
    123         *it.chain(_jaxpr.literals, flat_args),
    124         _jaxpr=_jaxpr.jaxpr,
   (...)
    128         **params,
    129     )
    130     return tree_util.tree_unflatten(out_tree(), outs)

File [/opt/conda/envs/genjax/lib/python3.11/site-packages/genjax/_src/core/interpreters/staging.py:124](http://localhost:8080/opt/conda/envs/genjax/lib/python3.11/site-packages/genjax/_src/core/interpreters/staging.py#line=123), in stage.<locals>.wrapped(*args, **kwargs)
    122 flat_fun, out_tree = api_util.flatten_fun_nokwargs(fun, in_tree)
    123 flat_avals = safe_map(get_shaped_aval, flat_args)
--> 124 typed_jaxpr = cached_stage_dynamic(flat_fun, tuple(flat_avals))
    125 return typed_jaxpr, (flat_args, in_tree, out_tree)

    [... skipping hidden 1 frame]

File [/opt/conda/envs/genjax/lib/python3.11/site-packages/genjax/_src/core/interpreters/staging.py:111](http://localhost:8080/opt/conda/envs/genjax/lib/python3.11/site-packages/genjax/_src/core/interpreters/staging.py#line=110), in cached_stage_dynamic(flat_fun, in_avals)
    109 @lu.cache
    110 def cached_stage_dynamic(flat_fun, in_avals):
--> 111     jaxpr, _, consts = pe.trace_to_jaxpr_dynamic(flat_fun, in_avals)
    112     typed_jaxpr = jc.ClosedJaxpr(jaxpr, consts)
    113     return typed_jaxpr

    [... skipping hidden 5 frame]

File [/opt/conda/envs/genjax/lib/python3.11/site-packages/genjax/_src/generative_functions/static.py:161](http://localhost:8080/opt/conda/envs/genjax/lib/python3.11/site-packages/genjax/_src/generative_functions/static.py#line=160), in _abstract_gen_fn_call(_, gen_fn, args)
    156 def _abstract_gen_fn_call(
    157     _: Address,
    158     gen_fn: GenerativeFunction,
    159     args: Tuple,
    160 ):
--> 161     return gen_fn.__abstract_call__(*args)

File [/opt/conda/envs/genjax/lib/python3.11/site-packages/genjax/_src/generative_functions/distributions/distribution.py:497](http://localhost:8080/opt/conda/envs/genjax/lib/python3.11/site-packages/genjax/_src/generative_functions/distributions/distribution.py#line=496), in ExactDensity.__abstract_call__(self, *args)
    495 def __abstract_call__(self, *args):
    496     key = jax.random.PRNGKey(0)
--> 497     return self.sample(key, *args)

File [/opt/conda/envs/genjax/lib/python3.11/site-packages/genjax/_src/generative_functions/distributions/distribution.py:573](http://localhost:8080/opt/conda/envs/genjax/lib/python3.11/site-packages/genjax/_src/generative_functions/distributions/distribution.py#line=572), in ExactDensityFromCallables.sample(self, key, *args)
    572 def sample(self, key, *args):
--> 573     return self.sampler(key, *args)

File [/opt/conda/envs/genjax/lib/python3.11/site-packages/genjax/_src/core/pytree.py:497](http://localhost:8080/opt/conda/envs/genjax/lib/python3.11/site-packages/genjax/_src/core/pytree.py#line=496), in Closure.__call__(self, *args, **kwargs)
    496 def __call__(self, *args, **kwargs):
--> 497     return self.fn(*self.dyn_args, *args, **kwargs)

File [/opt/conda/envs/genjax/lib/python3.11/site-packages/genjax/_src/generative_functions/distributions/tensorflow_probability/__init__.py:27](http://localhost:8080/opt/conda/envs/genjax/lib/python3.11/site-packages/genjax/_src/generative_functions/distributions/tensorflow_probability/__init__.py#line=26), in tfp_distribution.<locals>.sampler(key, *args, **kwargs)
     25 def sampler(key, *args, **kwargs):
     26     d = dist(*args, **kwargs)
---> 27     return d.sample(seed=key)

File [/opt/conda/envs/genjax/lib/python3.11/site-packages/tensorflow_probability/substrates/jax/distributions/distribution.py:1205](http://localhost:8080/opt/conda/envs/genjax/lib/python3.11/site-packages/tensorflow_probability/substrates/jax/distributions/distribution.py#line=1204), in Distribution.sample(self, sample_shape, seed, name, **kwargs)
   1190 """Generate samples of the specified shape.
   1191 
   1192 Note that a call to `sample()` without arguments will generate a single
   (...)
   1202   samples: a `Tensor` with prepended dimensions `sample_shape`.
   1203 """
   1204 with self._name_and_control_scope(name):
-> 1205   return self._call_sample_n(sample_shape, seed, **kwargs)

File [/opt/conda/envs/genjax/lib/python3.11/site-packages/tensorflow_probability/substrates/jax/distributions/distribution.py:1182](http://localhost:8080/opt/conda/envs/genjax/lib/python3.11/site-packages/tensorflow_probability/substrates/jax/distributions/distribution.py#line=1181), in Distribution._call_sample_n(self, sample_shape, seed, **kwargs)
   1178 sample_shape = ps.convert_to_shape_tensor(
   1179     ps.cast(sample_shape, tf.int32), name='sample_shape')
   1180 sample_shape, n = self._expand_sample_shape_to_vector(
   1181     sample_shape, 'sample_shape')
-> 1182 samples = self._sample_n(
   1183     n, seed=seed() if callable(seed) else seed, **kwargs)
   1184 samples = tf.nest.map_structure(
   1185     lambda x: tf.reshape(x, ps.concat([sample_shape, ps.shape(x)[1:]], 0)),
   1186     samples)
   1187 return self._set_sample_static_shape(samples, sample_shape, **kwargs)

File [/opt/conda/envs/genjax/lib/python3.11/site-packages/tensorflow_probability/substrates/jax/distributions/beta.py:293](http://localhost:8080/opt/conda/envs/genjax/lib/python3.11/site-packages/tensorflow_probability/substrates/jax/distributions/beta.py#line=292), in Beta._sample_n(self, n, seed)
    292 def _sample_n(self, n, seed=None):
--> 293   seed1, seed2 = samplers.split_seed(seed, salt='beta')
    294   concentration1 = tf.convert_to_tensor(self.concentration1)
    295   concentration0 = tf.convert_to_tensor(self.concentration0)

File [/opt/conda/envs/genjax/lib/python3.11/site-packages/tensorflow_probability/substrates/jax/internal/samplers.py:215](http://localhost:8080/opt/conda/envs/genjax/lib/python3.11/site-packages/tensorflow_probability/substrates/jax/internal/samplers.py#line=214), in split_seed(seed, n, salt, name)
    212   raise TypeError(
    213       '`n` must be a python `int` or an int Tensor, got {}'.format(repr(n)))
    214 with tf.name_scope(name or 'split_seed'):
--> 215   seed = sanitize_seed(seed, salt=salt)
    216   if JAX_MODE:
    217     from jax import random as jaxrand  # pylint: disable=g-import-not-at-top

File [/opt/conda/envs/genjax/lib/python3.11/site-packages/tensorflow_probability/substrates/jax/internal/samplers.py:143](http://localhost:8080/opt/conda/envs/genjax/lib/python3.11/site-packages/tensorflow_probability/substrates/jax/internal/samplers.py#line=142), in sanitize_seed(seed, salt, name)
    141 if salt is not None:
    142   salt = int(hashlib.sha512(str(salt).encode('utf-8')).hexdigest(), 16)
--> 143   seed = fold_in(seed, salt)
    145 if JAX_MODE:
    146   # Seed must be a jax.PRNGKey -- so just return it.
    147   return seed

File [/opt/conda/envs/genjax/lib/python3.11/site-packages/tensorflow_probability/substrates/jax/internal/samplers.py:179](http://localhost:8080/opt/conda/envs/genjax/lib/python3.11/site-packages/tensorflow_probability/substrates/jax/internal/samplers.py#line=178), in fold_in(seed, salt)
    176   from jax import random as jaxrand  # pylint: disable=g-import-not-at-top
    177   import jax.numpy as jnp  # pylint: disable=g-import-not-at-top
    178   return jaxrand.fold_in(
--> 179       seed, jnp.asarray(salt & np.uint32(2**32 - 1), dtype=SEED_DTYPE))
    180 if isinstance(salt, (six.integer_types)):
    181   seed = tf.bitwise.bitwise_xor(
    182       seed, np.uint64([salt & (2**64 - 1)]).view(np.int32))

OverflowError: Python int too large to convert to C long
```

This seems to be a TFP error somehow. I can get the same crash in the same environment with:
```
from tensorflow_probability.substrates import jax as tfp
key = jax.random.PRNGKey(17)
tfp.distributions.Beta(1, 1).sample(3, key)
```

This is with JAX version 0.4.31, TFP 0.23.0.

On an internal Google colab, which is using JAX version 0.4.32, TFP 0.25.0-dev, the TFP snippet runs fine. I'm asking about this internally.

**Comment by derifatives** [comment]
After some investigation, this is clearly a version mismatch. What I know so far:

The GenJAX setup seems to pull in TFP 0.23.0. [The latest stable release of TFP is 0.24.0.](https://github.com/tensorflow/probability/releases/tag/v0.24.0). It is not clear to me if 0.24.0 has this bug --- [Pavel checked in a fix for this only three weeks ago](https://github.com/tensorflow/probability/commit/8a5daf0224436fee1d7cd8e18382891ee92db3c4) (well after the 0.24.0), but it seems likely to me that this worked 0.24.0 + JAX 0.4.25. Unfortunately that combo is using numpy 1.

Most TFP users install the nightly, where this will be fixed. That may not be compatible with our build process, but wanted to mention it.

**Comment by derifatives** [comment]
Having discussed with the TFP team, we believe this was probably directly caused by GenJAX switching to numpy 2.0 (but we haven't proved this). The obvious fix for you is to figure out how to get a newer TFP. We can possibly make a new release for you if needed.

**Comment by sritchie** [comment]
We actually pinned to genjax <2.0.0 in genjax 0.5.1, want to verify that 0.5.1 fixes it?

**Comment by derifatives** [comment]
(I'm asuming you meant you pinned to numpy < 2.0.0 in GenJAX 0.5.1.)

Hmmm, I'm not sure your pin worked properly? I started a new conda environment a couple days ago for GenJAX 0.5.1, following the instructions on the install page, and it seems to have numpy 2.0.1. It is of course possible I misinstalled somehow.

I can confirm that this bug *does* happen in my install, with numpy 2.0.1 and GenJAX 0.5.1. I can try to reinstall if you think that's helpful.
